### PR TITLE
CCD-572 Fix for looping issue when fetching Az blob segments

### PR DESCRIPTION
**Pull request by dinesh1patel**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CCD-572

### Change description ###

1. Added mock test for empty list returned when fetching prefix from Az blob storage.
2. Added additional checks in while loop and conditional break. 
3. Confirmed successful grade build of project

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java [hunk 1]
```diff
@@ -71,19 +74,32 @@ public AzureImportAuditsClient(CloudBlobContainer cloudBlobContainer, Applicatio
         String currentDateTime;
         Integer azureImportAuditsGetLimit = applicationParams.getAzureImportAuditsGetLimit();
 
-        while (audits.size() < azureImportAuditsGetLimit) {
+        int counter = azureImportAuditsGetLimit;
```
Comment (MSancaktutar) [inline rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java:77]:
```
int maxDaysToCheck = 10 + azureImportAuditsGetLimit * 10; // Check 10 days at least + 10 days for each audit to collect.
       int daysChecked = 0;
```

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java [hunk 2]
```diff
@@ -71,19 +74,32 @@ public AzureImportAuditsClient(CloudBlobContainer cloudBlobContainer, Applicatio
...
+
+        while ((audits.size() < azureImportAuditsGetLimit) && counter > 0) {
```
Comment (MSancaktutar) [inline rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java:79]: while ((audits.size() < azureImportAuditsGetLimit) && daysChecked++ <= maxDaysToCheck )

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java [hunk 3]
```diff
@@ -71,19 +74,32 @@ public AzureImportAuditsClient(CloudBlobContainer cloudBlobContainer, Applicatio
...
+            if (blobsPage != null && !blobsPage.getResults().isEmpty()) {
+                List<ImportAudit> auditsLastBatch = populateListOfAudits(blobsPage);
+
+                audits.addAll(auditsLastBatch);
+            } else {
+                log.error("No result segment(s) found for prefix {} "
+                        + currentDateTime);
+            }
+
+            counter--;
```
Comment (MSancaktutar) [inline rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java:102]:
```
daysChecked++;
```

File: rest-api/src/test/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClientTest.java [hunk 4]
```diff
@@ -148,6 +150,17 @@ public void shouldFetchAllImportAuditsInCorrectDescOrder() throws Exception {
         verifyCloudBlockBlobBehaviour(b11, p11);
     }
 
+    @Test
+    public void shouldFetchNoImportAuditsWhenNoPrefixFound() throws Exception {
+
+        when(blobsPage1.getResults()).thenReturn(newArrayList());
+        when(blobsPage2.getResults()).thenReturn(newArrayList());
+        when(blobsPage3.getResults()).thenReturn(newArrayList());
+
```
[the remark this answers is not in the document]
Comment (MSancaktutar) [reply rest-api/src/test/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClientTest.java:162]: Hi Dinesh,

Please let's log 1 and only 1 error message if an only if we have found no files stored for a period of `maxDaysToCheck` days.

**Review by MSancaktutar — CHANGES_REQUESTED** [review]
A slight correction of logic required. Sorry for missing that in the previous chat.

File: rest-api/src/test/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClientTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dinesh1patel) [reply rest-api/src/test/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClientTest.java:165]: Completed, error will only be logged if the below condition is met:

if (audits.isEmpty() && (daysChecked == maxDaysToCheck)) {
    log.error("No import audits found over {} iterations", maxDaysToCheck);
}

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java [hunk 5]
```diff
@@ -71,19 +75,48 @@ public AzureImportAuditsClient(CloudBlobContainer cloudBlobContainer, Applicatio
...
+                List<ImportAudit> auditsLastBatch = populateListOfAudits(blobsPage);
+
+                audits.addAll(auditsLastBatch);
+            }
+
+            // cloudBlobContainer.listBlobsSegmented will return an empty list if prefixed value
+            // is not found, audits will remain at size 0.
+            // Need to break out of the while loop after processing max number of iterations.
+            if (counter == azureImportAuditsGetLimit) {
+                LOG.info("Exiting fetchLatestImportAudits, azureImportAuditsGetLimit:{}, "
```
[the remark this answers is not in the document]
Comment (MSancaktutar) [reply rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java]: As it is very unlikely, it is 99% an error to be in know of, either due to a data loss, or operational cleanup in a lower env, or a retrieval failure.

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java [hunk 6]
```diff
@@ -71,19 +74,37 @@ public AzureImportAuditsClient(CloudBlobContainer cloudBlobContainer, Applicatio
...
-            audits.addAll(auditsLastBatch);
+            if (blobsPage != null && !blobsPage.getResults().isEmpty()) {
+                List<ImportAudit> auditsLastBatch = populateListOfAudits(blobsPage);
+
+                audits.addAll(auditsLastBatch);
+            }
+
+            if (audits.isEmpty() && (daysChecked == maxDaysToCheck)) {
+                log.error("No import audits found over {} iterations", maxDaysToCheck);
+            }
```
Comment (MSancaktutar) [inline rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java:105]: This conditional log inside the loop just hinders smoother readability.
Could you please take it outside, after the loop?

**Review by MSancaktutar — CHANGES_REQUESTED** [review]
1 last comment to action, and it's good to go.

File: rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java (hunk 6 above)
Comment (dinesh1patel) [reply rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java]: Completed. Error log moved to outside of the while loop

Comment (dinesh1patel) [reply rest-api/src/main/java/uk/gov/hmcts/ccd/definition/store/rest/service/AzureImportAuditsClient.java]: Removed unused variable as flagged by Sonar scan

**Review by MSancaktutar — APPROVED** [review]

**Pull request merged**